java-checks/src/main/java/org/sonar/java/checks/spring/BeanMethodOfNonProxiedSingletonInvocationCheck.java

```diff
@@ -0,0 +1,121 @@
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    boolean isTargetClass = getConfigurationAnnotation((ClassTree) tree)
+      .map(BeanMethodOfNonProxiedSingletonInvocationCheck::hasProxyBeanMethodsDisabled)
+      .orElse(Boolean.FALSE);
+    if (!isTargetClass) {
+      return;
+    }
+    var visitor = new NonProxiedMethodInvocationVisitor((ClassTree) tree);
+    tree.accept(visitor);
+    visitor.locations.forEach(invocation -> reportIssue(invocation, "Replace this bean method invocation " + "with a dependency injection."));
+  }
+
+  private static Optional<AnnotationTree> getConfigurationAnnotation(ClassTree tree) {
+    SymbolMetadata metadata = tree.symbol().metadata();
+    for (SymbolMetadata.AnnotationInstance instance : metadata.annotations()) {
+      if (instance.symbol().type().is(CONFIGURATION_ANNOTATION)) {
+        return Optional.ofNullable(metadata.findAnnotationTree(instance));
+      }
+    }
+    return Optional.empty();
+  }
+
+
+  private static boolean hasProxyBeanMethodsDisabled(AnnotationTree annotation) {
+    return annotation.arguments().stream()
+      .filter(argument -> argument.is(Tree.Kind.ASSIGNMENT))
+      .map(AssignmentExpressionTree.class::cast)
+      .anyMatch(BeanMethodOfNonProxiedSingletonInvocationCheck::setsProxyBeanMethodsToFalse);
+  }
+
+  private static boolean setsProxyBeanMethodsToFalse(AssignmentExpressionTree assignment) {
+    return "proxyBeanMethods".equals(((IdentifierTree) assignment.variable()).name()) &&
+      Boolean.FALSE.equals(ExpressionsHelper.getConstantValueAsBoolean(assignment.expression()).value());
+  }
+
+  static class NonProxiedMethodInvocationVisitor extends BaseTreeVisitor {
+    private final ClassTree parentClass;
+    private final List<MethodInvocationTree> locations = new ArrayList<>();
+
+    NonProxiedMethodInvocationVisitor(ClassTree parentClass) {
+      this.parentClass = parentClass;
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      super.visitMethodInvocation(tree);
+      MethodTree declaration = tree.methodSymbol().declaration();
+      if (declaration == null) {
+        return;
+      }
+      if (returnsAPrototypeBean(declaration)) {
+        return;
+      }
+      Tree parent = declaration.parent();
+      if (parent == parentClass) {
+        locations.add(tree);
+      }
+    }
+
+    private static boolean returnsAPrototypeBean(MethodTree method) {
+      List<SymbolMetadata.AnnotationValue> annotationValues = method.symbol().metadata().valuesForAnnotation(SCOPE_ANNOTATION);
+      if (annotationValues == null) {
+        return false;
+      }
+      return annotationValues.stream()
+        .filter(argument -> List.of("value", "scopeName").contains(argument.name()))
+        .map(SymbolMetadata.AnnotationValue::value)
+        .anyMatch("Prototype"::equals);
```

Comment:
_Could_ also be combined in one `return`, but maybe your version with two returns is more readable:

```suggestion
      return annotationValues != null && annotationValues.stream()
        .filter(argument -> List.of("value", "scopeName").contains(argument.name()))
        .map(SymbolMetadata.AnnotationValue::value)
        .anyMatch("Prototype"::equals);
```